File: 프로그래머스/lv3/43105. 정수 삼각형/정수 삼각형.py
def solution(triangle):
    row = len(triangle)
    array = [[0 for _ in range(row)] for _ in range(row)]

    answer = 0
    array[0][0] = triangle[0][0]
    
    for i in range(0,row-1,1):
        for j in range(0,i+1,1):
            for d in range(2):
                array[i+1][j+d] = max(array[i][j]+triangle[i+1][j+d],array[i+1][j+d])
                
    for i in range(0,row,1):
        answer = max(answer,array[row-1][i])
    return answer

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.java
class Solution {
    public int solution(int[][] triangle) {
        int n = triangle.length;
        
        int [][] lst = new int[n+1][n+1];
        int [][] dp = new int[n+1][n+1];
        
        for(int i = 1; i <=n; i++ ){
            for(int j = 1; j<=i; j++){
                lst[i][j] = triangle[i-1][j-1];
            }
        }
        
        for(int i = 1; i <= n; i++){
            for(int j = 1; j <= i; j++){
                dp[i][j] = triangle[i-1][j-1] + Math.max(dp[i-1][j],dp[i-1][j-1]);
            }
        }
        
        int max = -1;
        for(int i=1; i<=n; i++){
            if(dp[n][i] > max){
                max = dp[n][i];
            }
        }
        
        return max;
    }
}